lightning/src/routing/router.rs

```diff
@@ -333,84 +607,280 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, targ
+			// There is still a chance of reaching them via last_hops though,
+			// so don't yet fail the payment here.
+			// If not, targets.pop() will not even let us enter the loop in step 2.
+			None => {},
+			Some(node) => {
+				add_entries_to_cheapest_to_target_node!(node, payee, 0);
+			},
+		}
 
-	for hop in last_hops.iter() {
-		let have_hop_src_in_graph =
-			if let Some(&(ref first_hop, ref features)) = first_hop_targets.get(&hop.src_node_id) {
-				// If this hop connects to a node with which we have a direct channel, ignore the
-				// network graph and add both the hop and our direct channel to the candidate set:
-				//
-				// Currently there are no channel-context features defined, so we are a
-				// bit lazy here. In the future, we should pull them out via our
-				// ChannelManager, but there's no reason to waste the space until we
-				// need them.
-				add_entry!(first_hop, *our_node_id , hop.src_node_id, dummy_directional_info, features.to_context(), 0);
-				true
-			} else {
-				// In any other case, only add the hop if the source is in the regular network
-				// graph:
-				network.get_nodes().get(&hop.src_node_id).is_some()
-			};
-		if have_hop_src_in_graph {
-			// BOLT 11 doesn't allow inclusion of features for the last hop hints, which
-			// really sucks, cause we're gonna need that eventually.
-			add_entry!(hop.short_channel_id, hop.src_node_id, target, hop, ChannelFeatures::empty(), 0);
+		// Step (1).
+		// If a caller provided us with last hops, add them to routing targets.
+		// Since this happens earlier than general path finding, they will be somewhat prioritized,
+		// although currently it matters only if the fees are exactly the same.
+		for hop in last_hops.iter() {
+			let have_hop_src_in_graph =
+				if let Some(&(ref first_hop, ref features, ref outbound_capacity_msat)) = first_hop_targets.get(&hop.src_node_id) {
+					// If this hop connects to a node with which we have a direct channel,
+					// ignore the network graph and add both the hop and
+					// our direct channel to the candidate set:
+					//
+					// Currently there are no channel-context features defined, so we are a
+					// bit lazy here. In the future, we should pull them out via our
+					// ChannelManager, but there's no reason to waste the space until we
+					// need them.
+					add_entry!(first_hop, *our_node_id , hop.src_node_id, dummy_directional_info, Some(outbound_capacity_msat / 1000), features.to_context(), 0);
+					true
+				} else {
+					// In any other case, only add the hop if the source is in the regular network
+					// graph:
+					network.get_nodes().get(&hop.src_node_id).is_some()
+				};
+			if have_hop_src_in_graph {
+				// BOLT 11 doesn't allow inclusion of features for the last hop hints, which
+				// really sucks, cause we're gonna need that eventually.
+				add_entry!(hop.short_channel_id, hop.src_node_id, payee, hop, None::<u64>, ChannelFeatures::empty(), 0);
```

Comment:
RIght, so this is where this field is used, so `htlc_maximum_msat` is not entirely unused (otherwise I'd drop the field).

Thus, your suggestion boils down to making them both `Option` in `RouteHint`?

Reply:
Oh, right, I confused it and the `None` being passed in for capacity_sats. Sure, just making htlc_minimum an Option would do it, though its confusing that the macro accesses that field directly.

Reply:
Since this macro uses both `RouteHint` and `DirectionalChannelInfo` for the same field, I'd have to update the field in both. It's weird to update `DirectionalChannelInfo.htlc_minimum_msat`, especially since it's not optional in `ChannelUpdate`. So `htlc_minimum_msat,` i think better stays a solid unit64 in `DirectionalChannelInfo`.

`htlc_maximum_msat`, on the other hand, is indeed optional in the current protocol.

So there's this difference.
There are several reasonable options now I think:
1. have both solid `u64` use u64::max() as a default value for `htlc_maximum_msat` in both `DirectionalChannelInfo` and `ChannelUpdate`
2. have both Optional in both `DirectionalChannelInfo` and `RouteHint`, and `DirectionalChannelInfo.htlc_minimum_msat` will just *always* be Some(xyz)
3. Fix the macro to consider that `RouteHint` is not the same as `DirectionalChannelInfo`.
4. Keep a mix of uint/Option as it is currently in my PR.

I personally don't see an issue with (4), it kinda reflects the spec as of today (although yeah, it's weird RouteHint has that struct).

Thus, perhaps (3) is the most reasonable option?

Reply:
Ended up using `DummyDirectionalChannelInfo`.